Lossy int to float coercion on `std.json.innerParseFromValue`

### Zig Version

0.15.0-dev.1425+7ee6dab39

### Steps to Reproduce and Observed Behavior

Currently facing this issue `type 'f64' cannot represent integer value` when trying to compile my project.

I believe that this is incorrect since the type I am using doesn't contain any floats. Currently I have only face this when trying to create a custom `jsonParse` or `jsonParseFromValue` in a struct.

Bellow is the stack trace of the error.

```bash
referenced by:
    jsonParseFromValue: src/main.zig:177:86
    refAllDecls__anon_440: /Users/raiden/.zvm/master/lib/std/testing.zig:1179:14
    test.simple test: src/main.zig:210:32
/Users/raiden/.zvm/master/lib/std/json/static.zig:570:44: error: type 'f64' cannot represent integer value '115792089237316195423570985008687907853269984665640564039457584007913129639935'
```

I believe that the potential bug was introduced in this commit https://github.com/ziglang/zig/pull/24632/commits/e664bf4d81e9266ee4749b5da88cab4554499bf6

I was able to create a way to reproduce this as you can check bellow

```bash
mkdir repro && cd repro
zig init
```

```zig
const std = @import("std");
const repro = @import("repro");

const Allocator = std.mem.Allocator;
const ParseFromValueError = std.json.ParseFromValueError;
const ParseError = std.json.ParseError;
const ParseOptions = std.json.ParseOptions;
const Value = std.json.Value;

/// Convert address to its representing bytes
pub fn addressToBytes(
    address: []const u8,
) error{ InvalidAddress, NoSpaceLeft, InvalidLength, InvalidCharacter }![20]u8 {
    const addr = if (std.mem.startsWith(u8, address, "0x")) address[2..] else address;

    if (addr.len != 40)
        return error.InvalidAddress;

    var addr_bytes: [20]u8 = undefined;
    _ = try std.fmt.hexToBytes(&addr_bytes, addr);

    return addr_bytes;
}

/// Convert a hash to its representing bytes
pub fn hashToBytes(
    hash: []const u8,
) error{ InvalidHash, NoSpaceLeft, InvalidLength, InvalidCharacter }![32]u8 {
    const hash_value = if (std.mem.startsWith(u8, hash, "0x")) hash[2..] else hash;

    if (hash_value.len != 64)
        return error.InvalidHash;

    var hash_bytes: [32]u8 = undefined;
    _ = try std.fmt.hexToBytes(&hash_bytes, hash_value);

    return hash_bytes;
}

/// Convert the struct fields into to a enum.
pub fn ConvertToEnum(comptime T: type) type {
    const info = @typeInfo(T);
    std.debug.assert(info == .@"struct");

    var enum_fields: [info.@"struct".fields.len]std.builtin.Type.EnumField = undefined;

    var count = 0;
    for (&enum_fields, info.@"struct".fields) |*enum_field, struct_field| {
        enum_field.* = .{
            .name = struct_field.name ++ "",
            .value = count,
        };
        count += 1;
    }

    return @Type(.{
        .@"enum" = .{
            .tag_type = usize,
            .fields = &enum_fields,
            .decls = &.{},
            .is_exhaustive = true,
        },
    });
}

pub const TokenExplorerTransaction = struct {
    /// The block number where the transaction was mined
    blockNumber: u64,
    /// The time when the transaction was committed.
    timeStamp: u64,
    /// The transaction hash
    hash: [32]u8,
    /// The transaction nonce.
    nonce: u64,
    /// The blockHash this transaction was mined.
    blockHash: [32]u8,
    /// The sender of this transaction
    from: [20]u8,
    /// The contract address in case it exists.
    contractAddress: [20]u8,
    /// The target address.
    to: [20]u8,
    /// The value sent. Only used for erc20 tokens.
    value: ?u256 = null,
    /// The token Id. Only used for erc721 and erc1155 tokens.
    tokenId: ?u256 = null,
    /// The token name.
    tokenName: []const u8,
    /// The token symbol.
    tokenSymbol: []const u8,
    /// The token decimal. Only used for erc20 and erc721 tokens.
    tokenDecimal: ?u8 = null,
    /// The index of this transaction on the mempool
    transactionIndex: usize,
    /// The gas limit of the transaction
    gas: u64,
    /// The gas price of this transaction.
    gasPrice: u64,
    /// The gas used by the transaction.
    gasUsed: u64,
    /// The cumulative gas used by the transaction.
    cumulativeGasUsed: u64,
    /// Input field that has been deprecated.
    input: []const u8 = "deprecated",
    /// The total number of confirmations
    confirmations: usize,

    pub fn jsonParse(
        allocator: Allocator,
        source: anytype,
        options: ParseOptions,
    ) ParseError(@TypeOf(source.*))!@This() {
        const json_value = try Value.jsonParse(allocator, source, options);
        return try jsonParseFromValue(allocator, json_value, options);
    }

    pub fn jsonParseFromValue(
        allocator: Allocator,
        source: Value,
        options: ParseOptions,
    ) ParseFromValueError!@This() {
        if (source != .object)
            return error.UnexpectedToken;

        var seen: std.enums.EnumFieldStruct(ConvertToEnum(@This()), u32, 0) = .{};
        var result: @This() = undefined;

        if (source.object.get("to")) |value| {
            if (value != .string)
                return error.UnexpectedToken;

            result.to = addressToBytes(value.string) catch return error.LengthMismatch;
            @field(seen, "to") = 1;
        }

        if (source.object.get("from")) |value| {
            if (value != .string)
                return error.UnexpectedToken;

            result.from = addressToBytes(value.string) catch return error.LengthMismatch;
            @field(seen, "from") = 1;
        }

        if (source.object.get("hash")) |value| {
            if (value != .string)
                return error.UnexpectedToken;

            result.hash = hashToBytes(value.string) catch return error.LengthMismatch;
            @field(seen, "hash") = 1;
        }

        if (source.object.get("blockHash")) |value| {
            if (value != .string)
                return error.UnexpectedToken;

            result.blockHash = hashToBytes(value.string) catch return error.LengthMismatch;
            @field(seen, "blockHash") = 1;
        }

        if (source.object.get("contractAddress")) |value| {
            if (value != .string)
                return error.UnexpectedToken;

            result.contractAddress = addressToBytes(value.string) catch return error.LengthMismatch;
            @field(seen, "contractAddress") = 1;
        }

        var iter = source.object.iterator();

        while (iter.next()) |key_value| {
            const field_name = key_value.key_ptr.*;

            inline for (std.meta.fields(@This())) |field| {
                if (std.mem.eql(u8, field.name, field_name)) {
                    if (@field(seen, field.name) == 0) {
                        @field(seen, field.name) = 1;
                        @field(result, field.name) = try std.json.innerParseFromValue(field.type, allocator, key_value.value_ptr.*, options);
                        break;
                    }
                }
            }
        }

        inline for (std.meta.fields(@This())) |field| {
            switch (@field(seen, field.name)) {
                0 => if (field.default_value_ptr) |default_value| {
                    @field(result, field.name) = @as(*const field.type, @ptrCast(@alignCast(default_value))).*;
                } else return error.MissingField,
                1 => {},
                else => {
                    switch (options.duplicate_field_behavior) {
                        .@"error" => return error.DuplicateField,
                        else => {},
                    }
                },
            }
        }

        return result;
    }
};

pub fn main() !void {
    // Prints to stderr, ignoring potential errors.
    std.debug.print("All your {s} are belong to us.\n", .{"codebase"});
    try repro.bufferedPrint();
}

test "simple test" {
    _ = std.testing.refAllDecls(TokenExplorerTransaction);
}
```

And then `zig build test -freference-trace=256` to run and get the error.

I am running this on a macOS with a M2 but this also happens if I set the `-Dtarget=x86_64-macos`

### Expected Behavior

To compile correctly without issues.

--- a/lib/std/json/static.zig
+++ b/lib/std/json/static.zig
@@ -567,8 +567,8 @@ pub fn innerParseFromValue(
             switch (source) {
                 .float => |f| {
                     if (@round(f) != f) return error.InvalidNumber;
-                    if (f > std.math.maxInt(T)) return error.Overflow;
-                    if (f < std.math.minInt(T)) return error.Overflow;
+                    if (f > @as(@TypeOf(f), @floatFromInt(std.math.maxInt(T)))) return error.Overflow;
+                    if (f < @as(@TypeOf(f), @floatFromInt(std.math.minInt(T)))) return error.Overflow;
                     return @as(T, @intFromFloat(f));
                 },
                 .integer => |i| {
@@ -770,7 +770,7 @@ fn sliceToInt(comptime T: type, slice: []const u8) !T {
     // Try to coerce a float to an integer.
     const float = try std.fmt.parseFloat(f128, slice);
     if (@round(float) != float) return error.InvalidNumber;
-    if (float > std.math.maxInt(T) or float < std.math.minInt(T)) return error.Overflow;
+    if (float > @as(f128, @floatFromInt(std.math.maxInt(T))) or float < @as(f128, @floatFromInt(std.math.minInt(T)))) return error.Overflow;
     return @as(T, @intCast(@as(i128, @intFromFloat(float))));
 }
 

